Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
-    return true;
-}
-
-
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          ResolveMOSIntersection
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Checks whether any of the Atoms in this AtomGroup are on top of
-//                  other MOSprite:s, and if so, attempt to move the OwnerMO of this out
-//                  so none of this' Atoms are inside any other MOS's silhouette anymore.
-
-bool AtomGroup::ResolveMOSIntersection(Vector &position, Matrix &rotation)
-{
-    if (!m_pOwnerMO->m_HitsMOs)
-        return true;
-
-	Vector atomOffset;
-	Vector atomPos;
-	Vector atomNormal;
-	Vector clearPos;
-	Vector exitDirection;
-	Vector atomExitVector;
-	Vector totalExitVector;
-    list<Atom *> intersectingAtoms;
-	MOID hitMOID = g_NoMOID;
-	MOID currentMOID = g_NoMOID;
-    MovableObject *pIntersectedMO = 0;
-    MOSRotating *pIntersectedMOS = 0;
-	float massA;
-	float massB;
-	float invMassA;
-	float invMassB;
-	float normMassA;
-	float normMassB;
-
-    hitMOID = g_NoMOID;
-    currentMOID = g_NoMOID;
-    pIntersectedMO = 0;
-    exitDirection.Reset();
-    atomExitVector .Reset();
-    totalExitVector.Reset();
-    intersectingAtoms.clear();
-
-    // First go through all atoms to find the first intersection and get the intersected MO
-    for (Atom *atom : m_Atoms)
-    {
-        atomOffset = atom->GetOffset().GetXFlipped(m_pOwnerMO->IsHFlipped());
-        atomOffset *= rotation;
-        atom->SetupPos(position + atomOffset);
-
-        atomPos = atom->GetCurrentPos();
-		hitMOID = g_SceneMan.GetMOIDPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
-		// Don't count MOIDs ignored
-		if (hitMOID != g_NoMOID && !atom->IsIgnoringMOID(hitMOID))
-		{
-			// Save the correct MOID to search for other atom intersections with
-			currentMOID = hitMOID;
-
-			// Get the MO we seem to be intersecting
-			MovableObject *tempMO = g_MovableMan.GetMOFromID(hitMOID);
-			RTEAssert(tempMO, "Intersected MOID couldn't be translated to a real MO!");
-			tempMO = tempMO->GetRootParent();
-
-			if (tempMO->GetsHitByMOs())
-			{
-				// Make that MO draw itself again in the MOID layer so we can find its true edges
-				pIntersectedMO = tempMO;
-				pIntersectedMO->Draw(g_SceneMan.GetMOIDBitmap(), Vector(), g_DrawMOID, true);
-				break;
-				// TODO: Remove
-				//                g_FrameMan.SaveBitmapToBMP(g_SceneMan.GetMOIDBitmap(), "MOIDMap");
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool AtomGroup::ResolveTerrainIntersection(Vector &position, unsigned char strongerThan) const {
+		std::list<Atom *> intersectingAtoms;
+		MOID hitMaterial = g_MaterialAir;
+
+		float strengthThreshold = (strongerThan != g_MaterialAir) ? g_SceneMan.GetMaterialFromID(strongerThan)->GetIntegrity() : 0.0F;
```

Comment:
Hmmm, we could probably just simplify this to `float strengthThreshold = g_SceneMan.GetMaterialFromID(strongerThan)->GetIntegrity()` but it'd probably be ever so slightly less efficient for air, which is probably where a lot of this is happening, so maybe we shouldn't.